### ISSUE TITLE
ToDo: Queue seek requests and invalidate deprecated requests

### ISSUE THREAD

**Issue by adipose** [clsid2/mpc-hc#856]
See : https://github.com/clsid2/mpc-hc/issues/835#issuecomment-763552202

**Comment by clsid2** [comment]
From linked topic:
> What exactly happens when we trigger multiple seeks in quick succession? My idea is to perform a seek. Then wait until it is fully done (if that is possible to track). If any new seeks are triggered in the meantime, then only store that new position value (overwriting existing queued value) and return. Then once the active seek is complete, perform the last queued seek. That should hopefully improve performance and stability. Because when people have crash of freeze issues it is often due to quick seeking.

**Comment by clsid2** [comment]
So as an experiment I added a sleep delay in CMainFrame::SeekTo() after SetPositions() call. I also added a bool parameter and an assert to check if there are simultaneous calls. It turns out that is not the case, at least not for seeks triggered from the seekbar. They are simply handled one after a time, so with a long delay added it will be busy for a while when processing several seeks. During that time the rest of the UI is blocked.

It is probably because the events are called through with window messages, which are handled one by one. Maybe it is possible to have some kind of look-ahead in the message handling? If seek action -> check if next queued message is also seek and if so skip first action -> check next message.

**Comment by adipose** [comment]
I don't think you can examine the message queue in that way, but we could easily create parallel data.  When adding a seek, we could put it in a global variable before sending the message.  If the message is not the same as the last seek, it might just ignore the message.

**Comment by clsid2** [comment]
Problem is that the interactions with the GUI are not processed immediately. So if I click twice, the second click is only processed after the first seek is complete.

**Comment by adipose** [comment]
https://github.com/clsid2/mpc-hc/pull/879

So, I basically did what I said, only I process all the messages, but the seeks get deferred/superseded if they start coming in too fast.  It's kind of a weird formula I used--if a new seek comes in less than 250ms after the prior one, it queues it instead of performing it.  Meanwhile, if more seeks keep coming in, they will replace the existing one, which means the timer keeps getting pushed out.

Then I put in a safety that if > 3 seeks are queued, it just processes the last one and kills the timer.  This prevents it from being deferred forever if many seeks are coming in.

No idea if this improves things, as my seeks don't really cause much issue.  But we can try it out on some machines that have issues.

**Comment by clsid2** [comment]
We can experiment with this after 1.9.9 stable release. I am probably going to release that soon.

**Comment by clsid2** [comment]
I made some modifications to your patch. It now delays a seek if it comes faster than 100ms after the last performed seek has completed (not started). This way it deals with slow seeks.
If another seek also needs to be delayed, then it overwrites the current queued one. Timer is not set to a fixed value, but set to 150ms since last performed seek. So delay can not accumulate. And "pause" between seeks is constant.
I also removed the delay checks in CheckScrollDistance of seekbar, since that does a similar thing as the new code.

Please try it out. It seems to be working pretty good here.

Test build:
https://www.sendspace.com/file/nwx7j7

**Comment by clsid2** [comment]
@ale5000-git @lifegpc
Please test as well

**Comment by adipose** [comment]
> I made some modifications to your patch. It now delays a seek if it comes faster than 100ms after the last performed seek has completed (not started). This way it deals with slow seeks.
> If another seek also needs to be delayed, then it overwrites the current queued one. Timer is not set to a fixed value, but set to 150ms since last performed seek. So delay can not accumulate. And "pause" between seeks is constant.
> I also removed the delay checks in CheckScrollDistance of seekbar, since that does a similar thing as the new code.
>

I didn't use less than 250ms because I couldn't get the seeks closer together than that. Maybe due to debug mode. I also wondered about using locks to prevent modifying the data while being processed.

Preventing delays from accumulating is smart. If they accumulate enough maybe processing them immediately is better than a timer?

**Comment by clsid2** [comment]
The existing CheckScrollDistance check already prevented seeks in close succession when using the seekbar. That may be why you couldn't get below 250ms. Using arrow keys to seek should trigger faster. 

Having a bit of time between seeks is actually intentional, because otherwise there is just a lot of wasteful work being done. Dragging the seekbar or holding down arrow button should give smooth and responsive behavior, while doing a minimal amount of seeks. I am probably going to lower current delay a bit for extra smoothness.

There doesn't seem to be simultaneous processing with the timer. If I add an artificial delay to the seek itself, then the playback time also stop updating while the seek is processed. That is also controlled with a timer. So that is good news and not a problem to worry about.

**Comment by adipose** [comment]
What I meant about immediate was this : suppose there are 10ms left in the 100ms delay. You could replace the timer, or you could just kill it and seek. Since timers aren't accurate beyond 10ms anyway, you might as well process that one, as it will probably be the next one to process anyway.

Timers are the lowest priority events so I wonder how they manage to sneak in between very rapid seeking. But glad it's working well.

**Comment by clsid2** [comment]
New test build:
https://www.sendspace.com/file/wdm8mf

**Comment by lifegpc** [comment]
I get crash when dragging the seek bar :(

https://user-images.githubusercontent.com/41434272/107132636-b2d0e480-691b-11eb-9ad2-e2b8ea3294dc.mp4

**Comment by lifegpc** [comment]
Seems only the video has very much key frame (1 key frame/s in average) have this issue. It seems not releated to the duration of the video.
I think it occurs because of these video need more time to seek.

**Comment by lifegpc** [comment]
test #879 , that patch works fine.

**Comment by clsid2** [comment]
Try this build:
https://www.sendspace.com/file/8lnsl6

**Comment by lifegpc** [comment]
seems works fine now.
